### ISSUE TITLE
G167: Using an adjacent button to label the purpose of a field

### ISSUE THREAD

**Issue by MichielBijl** [dequelabs/axe-core#1304]
According to [WCAG 2.1 Technique G167](https://www.w3.org/WAI/WCAG21/Techniques/general/G167), using an adjacent button to label a form field satisfies [WCAG 2.1 SC3.3.2](https://www.w3.org/TR/WCAG21/#labels-or-instructions).

Expectation: No issues are raised

Actual: aXe reports “Form elements must have labels”

Motivation: The provided code snippet is valid according to WCAG 2.0 / 2.1

<pre><code>
axe-core version: 3.6.0

Browser and Assistive Technology versions
Chrome 70.0.3538.110
</code></pre>

---

According to G167, aXe should pass this HTML…

```
<label for="s">Search</label>
<input type="search" id="s">
<button type="submit">Search</button>
```

…with this CSS:

```
label {
    position: absolute;
    clip: rect(0, 0, 0, 0);
}
```

You can check the [code on CodePen](https://codepen.io/Moiety/pen/1e061544e5f45f640c6dc0774c883300) or use this [debug link](https://s.codepen.io/Moiety/debug/1e061544e5f45f640c6dc0774c883300).

**Comment by WilcoFiers** [comment]
Hey Michiel. Thank you for reporting! This one took some digging. I had a difficult time reproducing it locally. Turns out, we already fixed this bug, but we haven't released the fix for it yet:
https://github.com/dequelabs/axe-core/pull/1187

the problem was with the hidden label, we were ignoring labels that are visually hidden, instead of ignoring labels that are hidden to assistive technologies. The next release of axe-core will have this fix in it. I'm going to close this as it has been resolved.

**Comment by MichielBijl** [comment]
Thanks Wilco 👍 

According to [Technique H44](https://www.w3.org/WAI/WCAG21/Techniques/html/H44#tests), an invisible label would fail [SC 3.3.2](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0#labels-or-instructions).

Not entirely sure how strict that is, as I would expect this to pass:

```HTML
<fieldset>
    <legend>Phone number </legend>

    <label for="area">Area code </label>
    <input type="text" id="area">

    <label for="phone-first">Middle three digits </label>
    <input type="text" id="phone-first">

    <label for="phone-last">Last four digits </label>
    <input type="text" id="phone-last">
</fieldset>
```

```CSS
label {
    position: absolute;
    clip: rect(0,0,0,0);
}
```

**Comment by WilcoFiers** [comment]
You're right. Form fields need to have a label as well as an accessible name. That label doesn't have to be a `<label>` element though, it can be a button or some text or an image or whatever. That's much harder to test automatically. Axe tests that the field has an accessible name, not that it has a visible label.

**Comment by MichielBijl** [comment]
Right, I was wondering how that could be automated. Thank you so much for your response!
